core/domain/exp_domain.py

```diff
@@ -2190,6 +2229,102 @@ def _convert_states_v27_dict_to_v28_dict(cls, states_dict):
+                        get_html_with_image_id_in_image_tag(
+                            interaction_feedback_html, image_counter,
+                            image_mapping))
+                state_dict['interaction']['default_outcome']['feedback'][
+                    'html'] = new_interaction_feedback_html
+
+            for answer_group_index, answer_group in enumerate(
+                    state_dict['interaction']['answer_groups']):
+                answer_group_html = answer_group['outcome']['feedback']['html']
+                (new_answer_group_html, image_counter, image_mapping) = (
+                    html_validation_service.get_html_with_image_id_in_image_tag(
+                        answer_group_html, image_counter, image_mapping))
+                state_dict['interaction']['answer_groups'][
+                    answer_group_index]['outcome']['feedback']['html'] = (
+                        new_answer_group_html)
+                if state_dict['interaction']['id'] == 'ItemSelectionInput':
+                    for rule_spec_index, rule_spec in enumerate(
+                            answer_group['rule_specs']):
+                        for x_index, x in enumerate(rule_spec['inputs']['x']):
+                            (
+                                new_x, image_counter, image_mapping) = (
+                                    html_validation_service.
+                                    get_html_with_image_id_in_image_tag(
+                                        x, image_counter, image_mapping))
+                            state_dict['interaction']['answer_groups'][
+                                answer_group_index]['rule_specs'][
+                                    rule_spec_index]['inputs']['x'][x_index] = (
+                                        new_x)
+
+            for hint_index, hint in enumerate(
+                    state_dict['interaction']['hints']):
+                hint_html = hint['hint_content']['html']
+                (new_hint_html, image_counter, image_mapping) = (
+                    html_validation_service.get_html_with_image_id_in_image_tag(
+                        hint_html, image_counter, image_mapping))
+                state_dict['interaction']['hints'][hint_index][
+                    'hint_content']['html'] = new_hint_html
+
+            if state_dict['interaction']['solution']:
+                solution_html = state_dict[
+                    'interaction']['solution']['explanation']['html']
+                (new_solution_html, image_counter, image_mapping) = (
+                    html_validation_service.
+                    get_html_with_image_id_in_image_tag(
+                        solution_html, image_counter, image_mapping))
+                state_dict['interaction']['solution']['explanation']['html'] = (
+                    new_solution_html)
+
+            if state_dict['interaction']['id'] in (
+                    'ItemSelectionInput', 'MultipleChoiceInput'):
```

Comment:
Make sure about the list of exploration which can have HTML content. (I think DranAndDrop Interaction also uses HTML as custom_args)